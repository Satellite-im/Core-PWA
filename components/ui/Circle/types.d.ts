export type CircleType = 'image' | 'icon' | 'random'
