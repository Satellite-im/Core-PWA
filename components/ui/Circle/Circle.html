<div 
  :class="`${square ? 'is-square' : 'is-rounded'} ${type === 'icon' ? 'circle-flex' : 'circle'}`"
  :style="`font-size: ${size * 0.55}px; height: ${size}px; width: ${size}px; background: ${color};`"
  @click="() => {$emit('click')}"
  >
  <img
    :src="source"
    v-if="type === 'image' && source"
    alt="Cirlce Image"
    :class="`${square ? 'is-square' : 'is-rounded'} image`">
  <jazzicon
    v-else-if="type === 'random'"
    class="jazzicon"
    :address="seed"
    :diameter="size"
  />
  <UiDynamicIcon
    size="1x"
    :icon="icon"
    v-else-if="icon" />
  <slot></slot>
</div>