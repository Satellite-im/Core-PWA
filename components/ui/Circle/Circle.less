.circle {
  &:extend(.font-secondary);
  &:extend(.no-select);
  display: inline-block;
  vertical-align: top;
  &-image {
    &:extend(.full-width);
    height: @full;
  }
}

.circle-flex {
  display: inline-flex;
  justify-content: center;
  align-items: center;
}

.is-rounded {
  border-radius: @half;
}
