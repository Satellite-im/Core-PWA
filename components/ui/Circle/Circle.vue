<template src="./Circle.html"></template>
<script lang="ts">
// @ts-ignore
import Jazzicon from 'vue-jazzicon'
import Vue, { PropType } from 'vue'

import { Icon } from '@/types/ui/icons'
import { CircleType } from './types'

export default Vue.extend({
  components: {
    Jazzicon,
  },
  props: {
    /**
     * Pixel diameter size of the circle
     */
    size: {
      type: Number,
      default: 45,
      required: false,
    },
    /**
     * Type of Circle, this determines what content we
     * load into the circle itself.
     */
    type: {
      type: String as PropType<CircleType>,
      default: 'random',
      required: true,
    },
    /**
     * Supported fontawesome icon
     */
    // eslint-disable-next-line vue/require-default-prop
    icon: {
      type: Object as PropType<Icon>,
      required: false,
    },
    /**
     * If provided the "circle" will have hard corners
     */
    square: {
      type: Boolean,
      default: false,
      required: false,
    },
    /**
     * Seed used to generate the random images
     */
    seed: {
      type: String,
      default: '0x0',
      required: false,
    },
    /**
     * Provided source if image icon type
     */
    source: {
      type: String,
      default: '',
      required: false,
    },
    /**
     * Color of the circles background
     */
    color: {
      type: String,
      default: '#16161e',
    },
  },
})
</script>
<style scoped lang="less" src="./Circle.less"></style>
