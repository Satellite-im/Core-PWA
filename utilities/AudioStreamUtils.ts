import { AudioState } from '~/store/audio/types'

const IS_TALKING_SENSITIVITY = 10 // %
const TALKING_OFF_DELAY = 500 // ms

export class AudioStreamUtils {
  stream: MediaStream
  audio: AudioState | null = null

  requestId: number | null = null
  level = 0
  isTalking = false

  constructor(stream: MediaStream, audio?: AudioState) {
    this.stream = stream
    this.audio = audio || null
  }

  public start() {
    this.listenToMicLevel()
  }

  public destroy() {
    this.stopListeningToMicLevel()
  }

  private listenToMicLevel = () => {
    if (!this.stream?.active) {
      return
    }
    const audioContext = new AudioContext()
    const gainNode = audioContext.createGain()
    const analyser = audioContext.createAnalyser()
    const microphone = audioContext.createMediaStreamSource(this.stream)
    const javascriptNode = audioContext.createScriptProcessor(2048, 1, 1)

    analyser.smoothingTimeConstant = 0

    microphone.connect(gainNode)
    gainNode.connect(analyser)
    analyser.connect(javascriptNode)
    javascriptNode.connect(audioContext.destination)

    const array = new Uint8Array(analyser.frequencyBinCount)

    const updateState = () => {
      if (!this.audio) {
        console.error('No audio state')
        return
      }
      // Update gain based on inputVolume
      gainNode.gain.setValueAtTime(
        this.audio.inputVolume / 100,
        audioContext.currentTime,
      )
      this.requestId = requestAnimationFrame(updateState)

      analyser.getByteFrequencyData(array)
      let values = 0

      const length = array.length
      for (let i = 0; i < length; i++) {
        values += array[i]
      }

      this.level = values / length
      this.updateIsTalking(this.level)
    }

    updateState()
  }

  private stopListeningToMicLevel() {
    if (this.requestId) {
      cancelAnimationFrame(this.requestId)
      this.requestId = null
    }
  }

  private updateIsTalking(micLevel: number) {
    const talking = Math.round(micLevel) > IS_TALKING_SENSITIVITY
    let talkingOffBuffer

    if (!talking && this.isTalking) {
      talkingOffBuffer = setTimeout(() => {
        this.isTalking = false
      }, TALKING_OFF_DELAY)
    } else if (talking && !this.isTalking) {
      this.isTalking = true

      if (talkingOffBuffer) {
        clearTimeout(talkingOffBuffer)
      }
    }
  }
}
