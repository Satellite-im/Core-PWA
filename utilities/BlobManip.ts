/**
 * @function blobToStream
 * @description convert Blob to ReadableStream. helpful for large uploads to avoid hitting the browser blob size limit
 * @param {Blob} file
 * @returns {ReadableStream} base64 thumbnail
 */
export function blobToStream(file: Blob): ReadableStream {
  // @ts-ignore
  const reader = file.stream().getReader()
  const stream = new ReadableStream({
    start(controller) {
      function push() {
        return reader
          .read()
          .then(({ done, value }: { done: boolean; value: Uint8Array }) => {
            if (done) {
              controller.close()
              return
            }
            controller.enqueue(value)
            push()
          })
      }
      push()
    },
  })
  return stream
}

/**
 * @method blobToBase64
 * @description convert File/Blob to base64 string
 * @param {Blob} file
 * @returns {Promise<string>} base64 thumbnail
 */
export function blobToBase64(file: Blob): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    const reader = new FileReader()
    reader.readAsDataURL(file)
    reader.onload = () => resolve(reader.result?.toString() || '')
    reader.onerror = (error) => reject(error)
  })
}
