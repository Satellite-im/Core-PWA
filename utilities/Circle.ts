export function addressToNumber(address: string) {
  return parseInt(address.slice(0, 10), 16)
}

export function toHex(str: string) {
  let res = ''
  for (let n = 0, l = str.length; n < l; n++) {
    const hex = Number(str.charCodeAt(n)).toString(16)
    res += hex
  }
  return res
}

export function stringToColor(str: string): string {
  const hash = stringToHash(str)
  return colors[hash]
}

function stringToHash(str: string): number {
  return (
    str.split('').reduce((p, c) => p + c.charCodeAt(0), 0) % (colors.length - 1)
  )
}

const colors = [
  '#fc5c65', // Fiery Rose
  '#fd9644', // Royal Orange
  '#fed330', // Sunglow
  '#26de81', // UFO Green
  '#2bcbba', // Maximum Blue Green
  '#eb3b5a', // Desire
  '#fa8231', // Princeton Orange
  '#f7b731', // Saffron
  '#20bf6b', // Green (Crayola)
  '#0fb9b1', // Tiffany Blue
  '#45aaf2', // Picton Blue
  '#4b7bec', // Blueberry
  '#a55eea', // Lavender Indigo
  '#778ca3', // Shadow Blue
  '#2d98da', // Bleu De France
  '#3867d6', // Royal Blue
  '#8854d0', // Amethyst
  '#a5b1c2', // Cadet Blue (Crayola)
  '#4b6584', // Dark Electric Blue
]
