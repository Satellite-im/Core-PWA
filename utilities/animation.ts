export type AnimationConfig = {
  timing: (v: number) => number
  draw: (v: number) => void
  duration: number
}

export function easeOutBack(v: number): number {
  const c1 = 1.2
  const c3 = c1 + 1

  return 1 + c3 * Math.pow(v - 1, 5) + c1 * Math.pow(v - 1, 4)
}

export function animate(config: AnimationConfig) {
  const { timing, draw, duration } = config
  const start = performance.now()

  const animate = (time: number) => {
    let timeFraction = (time - start) / duration
    if (timeFraction > 1) {
      timeFraction = 1
    }

    const progress = timing(timeFraction)
    draw(progress)

    if (timeFraction < 1) {
      requestAnimationFrame(animate)
    }
  }
  requestAnimationFrame(animate)
}

export function lerp(x: number, y: number, a: number) {
  return x * (1 - a) + y * a
}
