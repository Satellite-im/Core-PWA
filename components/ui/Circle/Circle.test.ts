import * as Circle from '~/utilities/Circle'

describe('Circle.addressToNumber', () => {
  test('0', () => {
    const result: any = Circle.addressToNumber('0.0.0.0')
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = Circle.addressToNumber('192.168.1.5')
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = Circle.addressToNumber('')
    expect(result).toMatchSnapshot()
  })
})

describe('Circle.stringToColor generated', () => {
  test('0', () => {
    const result: any = Circle.toHex('<?xml version="1.0" ?>\n<a b="c"/>\n')
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = Circle.toHex('<?xml version="1.0" ?><a b="c"/>')
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = Circle.toHex('')
    expect(result).toMatchSnapshot()
  })
})

describe('Circle.stringToColor', () => {
  it('should convert first did to color', () => {
    const argument: string =
      'did:key:z6MkiM2pazmVDL8BgUoRP1V3KU6tufrhMJbXQAEUD76ZVyZp '
    const result: string = Circle.stringToColor(argument)

    expect(result).toBe('#3867d6')
  })

  it('should convert second did to color', () => {
    const argument: string =
      'did:key:z6MkgJXKqJSredVSd6W21oXyjFjVaD5kBLfewRT4Gm6L5Nww'
    const result: string = Circle.stringToColor(argument)

    expect(result).toBe('#2bcbba')
  })
})
