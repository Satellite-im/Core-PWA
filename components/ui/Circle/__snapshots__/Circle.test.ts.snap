// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle.addressToNumber 0 1`] = `0`;

exports[`Circle.addressToNumber 1 1`] = `402`;

exports[`Circle.addressToNumber 2 1`] = `NaN`;

exports[`Circle.stringToColor 0 1`] = `"3c3f786d6c2076657273696f6e3d22312e3022203f3ea3c6120623d2263222f3ea"`;

exports[`Circle.stringToColor 1 1`] = `"3c3f786d6c2076657273696f6e3d22312e3022203f3e3c6120623d2263222f3e"`;

exports[`Circle.stringToColor 2 1`] = `""`;
